torch/utils/data/dataloader.py

```diff
@@ -288,8 +287,43 @@ def __init__(
         self.dataset = dataset
         self.num_workers = num_workers
         self.prefetch_factor = prefetch_factor
-        self.pin_memory = pin_memory
-        self.pin_memory_device = pin_memory_device
+
+        if pin_memory and pin_memory_device:
+            warnings.warn(
+                "pin_memory_device is deprecated, the current accelerator will be used as the device,"
+                f"ignore pin_memory_device='{pin_memory_device}'."
+            )
+        if pin_memory and not torch.accelerator.is_available():
+            warn_msg = (
+                "'pin_memory' argument is set as true but no accelerator is found, "
+                "then device pinned memory won't be used."
+            )
+            warnings.warn(warn_msg)
+
+        self.pin_memory = pin_memory and torch.accelerator.is_available()
+        self.pin_memory_device = (
+            acc.type
+            if self.pin_memory
+            and (acc := torch.accelerator.current_accelerator()) is not None
+            else None
+        )
+
+        # Currently, pin_memory would raise error on the MPS backend (see
+        # https://github.com/pytorch/pytorch/issues/86060), so forcibly
+        # disable pin_memory on MPS. Remove this restriction once pinned
+        # memory allocation for MPS is fixed.
+        if (
+            self.pin_memory
+            and (acc := torch.accelerator.current_accelerator()) is not None
+            and acc.type == "mps"
+        ):
```

Comment:
Can we replace this by 
```suggestion
        if (
            self.pin_memory_device is not None 
            and self.pin_memory_device == "mps"
        ):
```